src/quicktype-core/language/CPlusPlus.ts

```diff
@@ -971,99 +1036,209 @@ export class CPlusPlusRenderer extends ConvenienceRenderer {
+             */
+            propRecord.objectType = ObjectType.Class;
+            propRecord.kind = proposedIncludeKind !== undefined ? proposedIncludeKind : level === 0 ? IncludeKind.Include : IncludeKind.ForwardDeclare;
+        } else if (t instanceof EnumType) {
+            propRecord.objectType = ObjectType.Enum;
+            propRecord.kind = proposedIncludeKind !== undefined ? proposedIncludeKind : IncludeKind.ForwardDeclare;
+        } else if (t instanceof UnionType) {
+            /** Recurse into the union */
+            const [maybeNull, nonNulls] = removeNullFromUnion(t, true);
+            if (recurseIntoUnion) {
+                if (nonNulls !== undefined) {
+                    for (const tt of nonNulls) {
+                        /**
+                         * We can forward declare ANYTHING in a union, however
+                         * variants we need to include directly (they are typedefs)
+                         */
+                        this.updateIncludes(maybeNull === undefined, level + 1, includes, tt, defName);
+                    }
                 }
 
+                return;
+            }
+
+            /** 
+             * Even if forward declaration is possible, 
+             * but we don't want to check every possible union
+             * member types, simple include the definition.
+             */
+            propRecord.objectType = ObjectType.Union;
+            propRecord.kind = IncludeKind.Include;
+        }
+
+        if (!propertyTypes.has(typeName)) {
+            propertyTypes.set( typeName, propRecord);
+        } else {
+            /** 
+             * We don't care what we have already as a includekind
+             * for the type, we only update this if we want to
+             * set a stronger kind
+             */
+            if (propRecord.kind === IncludeKind.Include) {
+                propertyTypes.set( typeName, propRecord);
+            }
+        }
+    }
+
+    /** 
+     * if forwardDeclare is set to true we force it, level shows how 
+     * deep we are in in the definition
+     */
+    protected updateIncludes(forwardDeclare: boolean | undefined, level:number, includes: IncludeMap, propertyType: Type, defName:string): void {
+        /** God we don't really have custom sets */
+        let propertyTypes: Map<string, IncludeRecord> = new Map();
+
+        /** 
+         * Ok if the propertyType is an union type we can SURELY
+         * forward ANY directly reachable types
+         */
+        let proposedIncludeKind:IncludeKind | undefined = undefined;
+        if (forwardDeclare !== undefined) {
+            proposedIncludeKind = forwardDeclare ? IncludeKind.ForwardDeclare : undefined;
+        } else {
+            proposedIncludeKind = propertyType instanceof UnionType || propertyType instanceof ArrayType ? IncludeKind.ForwardDeclare : undefined;
+        }
+
+        /** Recurse into ArrayType */
+        if (propertyType instanceof ArrayType) {
```

Comment:
Why is this necessary for `ArrayType` and not for `MapType`, and why only once?  It could be an array of arrays.  (Not saying this is a bug, I just don't understand)

Reply:
This is practically a C++ thingie. The C++ compiler MUST know the exact definition of an "object" if it is a member. E.g. if you have:

class A
{
    B m_B;
}

in this case you MUST #include B's definition.
If B's included in a vector, variant, etc. you can clearly forward-declare it.
[This is ONLY true as we are using std::shared_ptr. If we were about to use std::optional
for optional types, we would be screwed, as optional on its own requires the definition of the
given class].
However further down the rabbithole (getting deeper in B's definition) I don't really care,Ű
e.g. it is just enough for me to know, whether of the "highest" level the given class's definition must be included or not. I don't need to include B's all child classes - as via pulling in B's definition automatically all the children's definition will be involved.

Reply:
So if it's `B m_B[]` then `B` must be included, but if it's `B m_B[][]`, then that's not necessary?